File: web-transport-tauri/src/rpc.ts
import { invoke } from "@tauri-apps/api/core";

export interface ConnectRequest {
	url: string;

	congestionControl?: "default" | "low-latency" | "throughput";
	// Only hex-encoded sha256 is currently supported.
	serverCertificateHashes?: string[];
}

export interface ConnectResponse {
	session: number;
}

export async function connect(payload: ConnectRequest): Promise<ConnectResponse> {
	return invoke<ConnectResponse>("plugin:web-transport|connect", {
		payload,
	});
}

export async function close(payload: CloseRequest): Promise<CloseResponse> {
	return invoke<CloseResponse>("plugin:web-transport|close", { payload });
}

export interface CloseRequest {
	session: number;
	code: number;
	reason: string | null;
}

export type CloseResponse = Record<string, never>;

export async function closed(payload: ClosedRequest): Promise<ClosedResponse> {
	return invoke<ClosedResponse>("plugin:web-transport|closed", { payload });
}

export interface ClosedRequest {
	session: number;
}

export type ClosedResponse = Record<string, never>;

export async function accept(payload: AcceptRequest): Promise<AcceptResponse> {
	return invoke<AcceptResponse>("plugin:web-transport|accept", { payload });
}

export interface AcceptRequest {
	session: number;
	bidirectional: boolean;
}

export interface AcceptResponse {
	stream: number;
}

export async function open(payload: OpenRequest): Promise<OpenResponse> {
	return invoke<OpenResponse>("plugin:web-transport|open", { payload });
}

export interface OpenRequest {
	session: number;
	bidirectional: boolean;
	sendOrder?: number;
}

export interface OpenResponse {
	stream: number;
}

export async function read(payload: ReadRequest): Promise<ReadResponse> {
	return invoke<ReadResponse>("plugin:web-transport|read", { payload });
}

export interface ReadRequest {
	session: number;
	stream: number;
}

export interface ReadResponse {
	data: Uint8Array | null;
}

export async function write(payload: WriteRequest): Promise<WriteResponse> {
	return invoke<WriteResponse>("plugin:web-transport|write", {
		payload,
	});
}

export interface WriteRequest {
	session: number;
	stream: number;
	data: Uint8Array | null;
}

export type WriteResponse = Record<string, never>;

export async function reset(payload: ResetRequest): Promise<ResetResponse> {
	return invoke<ResetResponse>("plugin:web-transport|reset", {
		payload,
	});
}

export interface ResetRequest {
	session: number;
	stream: number;
	code: number;
}

export type ResetResponse = Record<string, never>;

File: web-transport-tauri/src/rpc.rs
use bytes::Bytes;
use serde::{Deserialize, Serialize};
use tauri::Url;

#[derive(Debug, Deserialize, Serialize)]
#[serde(rename_all = "kebab-case")] // matches the WebTransport API
pub enum CongestionControl {
	Default,
	LowLatency,
	Throughput,
}

#[derive(Debug, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ConnectRequest {
	pub url: Url,
	pub congestion_control: Option<CongestionControl>,
	// Hex-encoded sha256 hashes.
	pub server_certificate_hashes: Option<Vec<String>>,
}

#[derive(Debug, Clone, Default, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ConnectResponse {
	pub session: usize,
}

#[derive(Debug, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct CloseRequest {
	pub session: usize,
	pub code: u32,
	pub reason: Option<String>,
}

#[derive(Debug, Clone, Default, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct CloseResponse {}

#[derive(Debug, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ClosedRequest {
	pub session: usize,
}

#[derive(Debug, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ClosedResponse {
	// TODO error code
}

#[derive(Debug, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct AcceptRequest {
	pub session: usize,
	pub bidirectional: bool,
}

#[derive(Debug, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct AcceptResponse {
	pub stream: usize,
}

#[derive(Debug, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct OpenRequest {
	pub session: usize,
	pub bidirectional: bool,
	pub send_order: Option<i32>,
}

#[derive(Debug, Clone, Default, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct OpenResponse {
	pub stream: usize,
}

#[derive(Debug, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ReadRequest {
	pub session: usize,
	pub stream: usize,
}

#[derive(Debug, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ReadResponse {
	// If None, then the stream is closed.
	pub data: Option<Bytes>,
}

#[derive(Debug, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct WriteRequest {
	pub session: usize,
	pub stream: usize,
	pub data: Bytes,
}

#[derive(Debug, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct WriteResponse {}

#[derive(Debug, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct FinishRequest {
	pub session: usize,
	pub stream: usize,
}

#[derive(Debug, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct FinishResponse {}

#[derive(Debug, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ResetRequest {
	pub session: usize,
	pub stream: usize,
	pub code: u32,
}

#[derive(Debug, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ResetResponse {}
